backend/benefit/applications/benefit_aggregation.py

```diff
@@ -0,0 +1,168 @@
+import operator
+from dataclasses import dataclass, field
+from decimal import Decimal
+from itertools import chain
+from typing import List, Optional
+
+from applications.enums import ApplicationStatus
+from applications.models import Application
+from calculator.models import PreviousBenefit
+from common.utils import date_range_overlap, duration_in_months, pairwise
+from dateutil.relativedelta import relativedelta
+from django.utils.text import format_lazy
+from django.utils.translation import gettext_lazy as _
+
+
+@dataclass
+class PastBenefitInfo:
+    validation_errors: List[str] = field(default_factory=list)
+    validation_messages: List[str] = field(default_factory=list)
+    months_used: Optional[
+        Decimal
+    ] = None  # None is used if employee info is not entered yet
+    months_remaining: Optional[Decimal] = None  # None means that there is no limit
+
+
+BENEFIT_MAX_DURATION = Decimal("12")
+APPRENTICESHIP_BENEFIT_MAX_DURATION = Decimal("12")
+BENEFIT_WAITING_PERIOD_MONTHS = 24
+
+
+def get_past_benefit_info(
+    application,
+    company,
+    social_security_number,
+    start_date,
+    end_date,
+    apprenticeship_program,
+):
+    # the application field values are separate parameters, because validation is done
+    # before assigning values, and if a new Application is being created, an Application doesn't yet exist when
+    # the rest framework does the validation.
+    # The Application parameter is only used to ensure that the application being validated isn't validated
+    # against itself.
+
+    past_benefit_info = PastBenefitInfo()
+
+    if not social_security_number or not end_date:
+        # the employee info hasn't been entered yet
+        return past_benefit_info
+
+    recent_benefits = _get_benefits_relevant_for_validation(
+        _get_past_benefits(application, company, social_security_number, end_date),
+        start_date,
+    )
+
+    past_benefit_info.validation_errors.extend(
+        _get_benefit_overlap_errors(recent_benefits, start_date, end_date)
+    )
+
+    past_benefit_info.months_used = sum(
+        benefit.duration_in_months for benefit in recent_benefits
+    )
+    if apprenticeship_program:
+        # Kanslia Helsinki-lisä Teams discussion 2021-12-09: there's no upper limit defined for
+        # sequentially granted apprenticeship benefits
+        past_benefit_info.months_remaining = None
+    else:
+        past_benefit_info.months_remaining = max(
+            Decimal(0), BENEFIT_MAX_DURATION - past_benefit_info.months_used
+        )
+
+    if (
+        not apprenticeship_program
+        and past_benefit_info.months_used + duration_in_months(start_date, end_date)
+        > Application.BENEFIT_MAX_MONTHS
```

Comment:
What's the difference between BENEFIT_MAX_DURATION and BENEFIT_MAX_MONTHS? They're both 12 at the moment.

The code is ok as it is here and couldn't the if-clause also be done like this:
`
if(not apprenticeship_program and duration_in_months(start_date, end_date) > past_benefit_info.months_remaining)
`

Reply:
Good point, need to fix

Reply:
fixed